Add UniformBinning to schemav2

eguiraud opened the pull request:
When defining a Binning, the `edges` attribute can now be either a list of floats (like before) or an instance of UniformBinning(n, low, high). When using the latter option, the Binning correction is optimized to use a simple arithmetic formula rather than binary search to look up the bin index.

This is a fresh (better) take on #181 .

Example usage:

```python
from correctionlib import schemav2 as schema

if __name__ == "__main__":
    corr = schema.Correction(
        name="test",
        version=2,
        inputs=[schema.Variable(name="x", type="real")],
        output=schema.Variable(name="a scale", type="real"),
        data=schema.Binning(
            nodetype="binning",
            input="x",
            edges=schema.UniformBinning(n=2, low=0., high=2.),
            content=[1.0, 2.0],
            flow="error",
        ),
    )
    e = corr.to_evaluator()
    print(e.evaluate(0.5))
    print(e.evaluate(1.0))
    print(e.evaluate(3.0)) # errors out
```

To do:
- [x] add tests
- [x] add support for MultiBinning

nsmith- commented on src/correction.cc:
```diff
@@ -319,17 +319,35 @@ double HashPRNG::evaluate(const std::vector<Variable::Type>& values) const {
+    }
+    _NonUniformBins bins{std::move(edges)};
+    bins_ = std::move(bins);
+  } else { // must be UniformBinning
```
`else if ( edgesObj.IsObject() )`
then a final else that throws an error

eguiraud replied:
done!

nsmith- commented on src/correction.cc:
```diff
@@ -319,17 +319,35 @@ double HashPRNG::evaluate(const std::vector<Variable::Type>& values) const {
+    bins_ = std::move(bins);
+  } else { // must be UniformBinning
+    const JSONObject uniformBins{edgesObj.GetObject()};
+    const auto n = uniformBins.getRequired<uint64_t>("n");
```
Check for n=0 here (since we don't always run the json through pydantic)

eguiraud replied:
done!

nsmith- reviewed (changes requested):
Yeah this is the right approach I think.
Looks good, some small extra validation comments only

eguiraud commented:
Addressed review comments, added some tests for Binning+UniformBinning, fixed a bug in an edge case that the test found.

I will work on adding support for MultiBinning as well as soon as possible (could also be a separate PR).

nsmith- reviewed (approved):
Looks great! Thanks so much for this addition, and sorry I didn't review it sooner
